---
title: "Ecological Genomics"
output:
  rmdformats::readthedown:
    highlight: kate
    lightbox: true
    gallery: true
    # toc: 2
---


```{r setup, echo=FALSE, cache=FALSE}
library(knitr)
library(rmdformats)
library(DT)

## Global options
options(max.print="75")
opts_chunk$set(echo=FALSE,
	             cache=TRUE,
               prompt=FALSE,
               tidy=TRUE,
               comment=NA,
               message=FALSE,
               warning=FALSE)
opts_knit$set(width=75)
```

**Affiliation**:  Keller Lab; Department of Plant Biology, University of Vermont  
**E-mail contact**:  anoob.prakash@uvm.edu  
**Start Date**: 2020-01-13  
**End Date**: 2020-05-08  
**Project Descriptions**: Climate change adaptation in red spruce  

<!-- ## Table of Contents:   -->
<!-- [Working with the terminal ](#terminal)   -->
<!-- [bash commands](#basicBash)  -->
<!-- [Working with fastq](./fastq.html)    -->
<!-- [Trimmomatic]()   -->




<!-- [Course Materials](https://docs.google.com/document/d/1MuXPS3hNg3wIxsKw-jQmKY7y57Ka2qqZwHYKRAbpNMw/edit?usp=sharing)  -->

<!-- [UNIX cheatsheet](https://files.fosswire.com/2007/08/fwunixref.pdf) -->

-----
# Working with the terminal <a name="terminal"></a>  
**Log into UNIX server through RStudio terminal**  

Update: No need to use the ssh package, just go straight to the terminal and start typing.  
require(ssh)
session <- ssh_connect("ap1@pbio381.uvm.edu")
print(session)

Start a Rmarkdown file. Go to terminal on RStudio, type in the following to get root access:  

`/usr/bin/env bash -l`  

type in your password  

You have root access now:
 type in the details to login to the server  
 
 `ssh servername`
 
 For this course, type in netID followed by the server ID. For example:
 
 `ssh ap1@pbio381.uvm.edu`  

Next type in your net ID password, and now you can accesss the UNIX shell like you would be accesssing through the terminal on MAC or through PUTTY on Windows.

If you want run the codes that you currently have Rmarkdown, use `Ctrl + Shift + Enter`

Additonal tip for Windows users, there is an alterantive to putty on windows store. Just download and install ubuntu and it works as a linux shell. It seems more integrated and part of the windows system, however you cannot save login credentials like you can on putty.  
   
 
 More details on RStudio terminal can be  found here: https://support.rstudio.com/hc/en-us/articles/115010737148-Using-the-RStudio-Terminal   
 
**Location of shared data on the server:** `/data/project_data/RS_ExomeSeq`  
 
*** 
 
## bash commands <a name="basicBash"></a>  

Description                                     | Commands  
----------------------------------------------- | --------
Finding what directory you’re in:               |   `pwd`  
Listing files in your current directory:        |   `ll`  `ls`
Changing to a new directory:                    |   `cd`  
Making a new folder                             |`mkdir` *foldername*  
Copying or moving files from one location to another| `cp` *filename destinationpath/*,  `mv` *filename destinationpath/*  
Peeking into the first or last few lines of a file  | `head` *filename*, `tail` *filename*  
Searching within a file for a match              | `grep` **search string** *filename*  
Outputing the results of a command to a new file | `grep` **search string** *filename* >outputfilename  
Using the “pipe” to send the output of one command to the input of another | `grep` **text
Removing files or folders [use with caution as the action is irreversible]  | `rm`   
<p style="color:green"> Github specific</p>  
Description                                     | Commands  
----------------------------------------------- | --------
Git add                                         | `git add`  
Pull from the server                            | `git pull`  
Commit changes                                  | `git commit`
Push to the server                              | `git push`  
Cloning a git repo                              |    `git clone`


## File transfer protocol  
Files can be transfered from the server to the local machine through pushing it to the github. However, when the files are too large to commit a push to the GitHub server, a file transfer software like [WinSCP](https://winscp.net/eng/index.php) is required ( [FileZilla](https://filezilla-project.org/?AFFILIATE=6732&__c=1) or [Cyberduck](https://cyberduck.io/) for Mac). Such transfer clients helps to establish a secure connection and carry out file transfer between a local and a remote computer. Just the server credentials are required to start using the software. Make sure the connection type is SFTP, which stands for SSH File Transfer Protocol, or Secure File Transfer Protocol.  


***
# Course Materials amd Notes  
## [Sequencing methods](https://docs.google.com/document/d/1MuXPS3hNg3wIxsKw-jQmKY7y57Ka2qqZwHYKRAbpNMw/edit?usp=sharing)
[Tutorial on bash](https://pespenilab.github.io/Ecological-Genomics/Tutorial/2020-01-22_Command_Line_Unix.html)   

### [Course website](https://pespenilab.github.io/Ecological-Genomics/)  

***  
